dbms/src/TiDB/OwnerManager.cpp

```diff
@@ -0,0 +1,520 @@
+    }
+    case State::Init:
+    case State::CancelDone:
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "unexpected state: {}", magic_enum::enum_name(state));
+    }
+    return {run_next_campaign, old_session};
+}
+
+void EtcdOwnerManager::tryChangeState(State coming_state)
+{
+    std::unique_lock lk(mtx_camaign);
+    // The campaign is stopping, it will cause
+    // - leader key deleted in watch thread
+    // - etcd lease expired in keepalive task
+    // Do not overwrite `CancelByStop` because the next campaign is expected to be stopped.
+    if (state != State::CancelByStop)
+    {
+        // ok to set the state
+        state = coming_state;
+    }
+}
+
+void EtcdOwnerManager::camaignLoop(Etcd::SessionPtr session)
+{
+    try
+    {
+        while (true)
+        {
+            bool run_next_campaign = true;
+            Etcd::SessionPtr old_session = session; // workaround for clang-tidy
+            std::tie(run_next_campaign, session) = runNextCampaign(std::move(old_session));
+            if (!run_next_campaign)
+                break;
+
+            const auto lease_id = session->leaseID();
+            LOG_DEBUG(log, "new campaign loop with lease_id={:x}", lease_id);
+            Etcd::LeaderKey new_leader;
+            grpc::Status status;
+            std::tie(new_leader, status) = client->campaign(campaign_name, id, lease_id);
+            if (!status.ok())
+            {
+                // if error, continue next campaign
+                LOG_INFO(
+                    log,
+                    "failed to campaign, id={} lease={:x} code={} msg={}",
+                    id,
+                    lease_id,
+                    status.error_code(),
+                    status.error_message());
+                continue;
+            }
+
+            auto owner_key = getOwnerKey(id);
+            if (!owner_key)
+            {
+                // if error, continue
+                continue;
+            }
+
+            // become owner
+            toBeOwner(std::move(new_leader));
+            LOG_INFO(log, "become the owner with lease={:x}", lease_id);
+
+            grpc::ClientContext watch_ctx;
+            // waits until owner key get expired and deleted
+            th_watch_owner = ThreadFactory::newThread(
+                false,
+                "OwnerWatch",
+                [this, key = owner_key.value(), &watch_ctx] {
+                    watchOwner(key, &watch_ctx);
+                });
+
+            {
+                // etcd session expired / owner key get deleted / caller cancel,
+                // it means this node is not owner anymore
+                std::unique_lock lk(mtx_camaign);
+                cv_camaign.wait(lk, [this] { return state != State::Normal; });
+                LOG_INFO(log, "{}", magic_enum::enum_name(state));
+            }
+
+            watch_ctx.TryCancel();
+            th_watch_owner.join();
+            retireOwner();
+
+            LOG_WARNING(log, "is not the owner");
```

Comment:
ditto

Reply:
I've added the owner id to the logger as a suffix. So is this good enough for the logging?
```
[2023/02/27 10:46:49.672 +08:00] [INFO] [OwnerManager.cpp:150] ["start campaign owner"] [source="owner_id=owner_0"] [thread_id=1]
[2023/02/27 10:46:49.675 +08:00] [INFO] [OwnerManager.cpp:273] ["become the owner with lease=5fea8665748d864b"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.675 +08:00] [INFO] [gtest_owner_manager.cpp:256] ["test wait for n*ttl"] [thread_id=1]
[2023/02/27 10:46:49.676 +08:00] [WARN] [OwnerManager.cpp:295] ["is not the owner"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.676 +08:00] [WARN] [OwnerManager.cpp:169] ["state change by failpoint CancelByKeyDeleted -> CancelByLeaseInvalid"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.676 +08:00] [INFO] [OwnerManager.cpp:200] ["etcd session is expired, create a new one"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:51.681 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:53.683 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:55.685 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:57.686 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.687 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.765 +08:00] [INFO] [OwnerManager.cpp:273] ["become the owner with lease=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:47:49.675 +08:00] [INFO] [gtest_owner_manager.cpp:270] ["test wait for n*ttl passed"] [thread_id=1]
[2023/02/27 10:47:49.676 +08:00] [INFO] [OwnerManager.cpp:347] ["watcher is closed"] [source="owner_id=owner_0"] [thread_id=46]
[2023/02/27 10:47:49.677 +08:00] [WARN] [OwnerManager.cpp:295] ["is not the owner"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:47:49.678 +08:00] [INFO] [OwnerManager.cpp:184] ["break campaign loop, disabled"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:47:49.679 +08:00] [INFO] [OwnerManager.cpp:488] ["revoke session, code=0 msg="] [source="owner_id=owner_0"] [thread_id=43]
```

<details>

<summary> trace level log </summary>

```
[2023/02/27 10:46:49.671 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:46:59"] [source="lease=5fea8665748d864b"] [thread_id=1]
[2023/02/27 10:46:49.672 +08:00] [INFO] [OwnerManager.cpp:150] ["start campaign owner"] [source="owner_id=owner_0"] [thread_id=1]
[2023/02/27 10:46:49.673 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:46:59"] [source="lease=5fea8665748d864b"] [thread_id=42]
[2023/02/27 10:46:49.673 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d864b"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.675 +08:00] [INFO] [OwnerManager.cpp:273] ["become the owner with lease=5fea8665748d864b"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.675 +08:00] [INFO] [gtest_owner_manager.cpp:256] ["test wait for n*ttl"] [thread_id=1]
[2023/02/27 10:46:49.676 +08:00] [DEBUG] [OwnerManager.cpp:335] ["watch finish, code=1 msg=CANCELLED"] [source="owner_id=owner_0"] [thread_id=44]
[2023/02/27 10:46:49.676 +08:00] [WARN] [OwnerManager.cpp:295] ["is not the owner"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.676 +08:00] [WARN] [OwnerManager.cpp:169] ["state change by failpoint CancelByKeyDeleted -> CancelByLeaseInvalid"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.676 +08:00] [INFO] [OwnerManager.cpp:200] ["etcd session is expired, create a new one"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.680 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:46:59"] [source="lease=5fea8665748d8650"] [thread_id=43]
[2023/02/27 10:46:49.680 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:49.680 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:46:59"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:46:51.681 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:51.681 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:53.683 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:53.683 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:55.685 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:55.685 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:56.347 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:06"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:46:56.685 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:06"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:46:57.686 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:57.686 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.687 +08:00] [INFO] [OwnerManager.cpp:260] ["failed to campaign, id=owner_0 lease=5fea8665748d8650 code=4 msg=Deadline Exceeded"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.687 +08:00] [DEBUG] [OwnerManager.cpp:247] ["new campaign loop with lease_id=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.765 +08:00] [INFO] [OwnerManager.cpp:273] ["become the owner with lease=5fea8665748d8650"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:46:59.786 +08:00] [TRACE] [OwnerManager.cpp:344] ["watch key=/tiflash/s3gc/owner/5fea8665748d8650 ok=true resp=header { cluster_id: 3896301785396801271 member_id: 1909762081666129898 revision: 1205532 raft_term: 4 } created: true"] [source="owner_id=owner_0"] [thread_id=46]
[2023/02/27 10:47:03.352 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:13"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:03.468 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:13"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:10.135 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:20"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:10.522 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:20"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:17.189 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:27"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:24.093 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:34"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:24.116 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:34"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:31.019 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:41"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:31.357 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:41"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:38.512 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:48"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:38.758 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:48"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:45.737 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:55"] [source="lease=5fea8665748d8650"] [thread_id=45]
[2023/02/27 10:47:46.173 +08:00] [DEBUG] [Client.cpp:266] ["keep alive update deadline, ttl=10 lease_deadline=2023-02-27 10:47:56"] [source="lease=5fea8665748d8650"] [thread_id=42]
[2023/02/27 10:47:49.675 +08:00] [INFO] [gtest_owner_manager.cpp:270] ["test wait for n*ttl passed"] [thread_id=1]
[2023/02/27 10:47:49.676 +08:00] [TRACE] [OwnerManager.cpp:344] ["watch key=/tiflash/s3gc/owner/5fea8665748d8650 ok=false resp="] [source="owner_id=owner_0"] [thread_id=46]
[2023/02/27 10:47:49.676 +08:00] [INFO] [OwnerManager.cpp:347] ["watcher is closed"] [source="owner_id=owner_0"] [thread_id=46]
[2023/02/27 10:47:49.676 +08:00] [DEBUG] [OwnerManager.cpp:366] ["watch finish, code=1 msg=CANCELLED"] [source="owner_id=owner_0"] [thread_id=46]
[2023/02/27 10:47:49.677 +08:00] [WARN] [OwnerManager.cpp:295] ["is not the owner"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:47:49.678 +08:00] [INFO] [OwnerManager.cpp:184] ["break campaign loop, disabled"] [source="owner_id=owner_0"] [thread_id=43]
[2023/02/27 10:47:49.679 +08:00] [INFO] [OwnerManager.cpp:488] ["revoke session, code=0 msg="] [source="owner_id=owner_0"] [thread_id=43]
```
</details>

Reply:
OK. In cloud env, the log of different tiflash is mix together, so maybe log out some instance information may help, like `xxx.xxx.xxx.xxx:xxx is not the owner`, but now it also acceptable for me.

Reply:
We use the `raft_config.flash_server_addr` as owner_id, so the log with show "[is not the owner] [source="x.x.x.x:x"]"
https://github.com/pingcap/tiflash/blob/bf82b38a8255acba4bc5621eb2c4d9e01c9ee824/dbms/src/Storages/Transaction/TMTContext.cpp#L93-L97